### PR TITLE
Implement no_grad functionality following PyTorch API

### PR DESCRIPTION
**Pull request by yajm**
This PR adds gradient control functionality to MLX that matches PyTorch's no_grad, enable_grad, and set_grad_enabled APIs.

## Summary
- **C++ Backend**: Thread-local gradient state with RAII guards
- **Python Frontend**: Context managers and decorators matching PyTorch API  
- **Integration**: Existing transforms (vjp, jvp) respect gradient state
- **Backward Compatible**: All existing code continues to work

## Features Added
- `no_grad()` context manager and decorator
- `enable_grad()` context manager and decorator  
- `set_grad_enabled(mode)` conditional gradient control
- `mx.is_grad_enabled()` and `mx.set_grad_enabled()` core functions

## Testing
- ✅ All existing tests pass (verified backward compatibility)
- ✅ New C++ tests: 27 assertions covering gradient state control
- ✅ New Python tests: Context managers, decorators, nested scenarios
- ✅ Manual verification with working test program

## Performance
- Zero overhead when gradients enabled (default behavior)
- Significant memory savings when gradients disabled during inference

## Usage
```python
import mlx.core as mx
from mlx.autograd import no_grad, enable_grad, set_grad_enabled

# Context manager
with no_grad():
    output = model(input)  # No gradients computed

# Decorator  
@no_grad()
def inference(x):
    return model(x)

# Conditional gradients
with set_grad_enabled(is_training):
    loss = compute_loss(predictions, targets)
```

This implementation follows MLX's architecture patterns and provides the exact same API as PyTorch for easy migration.

### CODE REVIEW

**Comment by awni** [comment]
This kind of functionality is not necessary in MLX. If you don't take the gradient of a function you won't compute gradients.. it's really as simple as that:

```
# Context manager
#. with no_grad(): <-- You don't need that
output = model(input)  # No gradients computed by default
```

If you want gradients you have to ask for them using `mx.grad` or similar functionality.
